src/state.rs

```diff
@@ -224,149 +263,126 @@ impl State {
-    //     {
-    //         Ok(cid) => cid,
-    //         Err(err) => abort!(USR_SERIALIZATION, "failed to store initial state: {:}", err),
-    //     };
-    //     if let Err(err) = sdk::sself::set_root(&cid) {
-    //         abort!(USR_ILLEGAL_STATE, "failed to set root ciid: {:}", err);
-    //     }
-    //     cid
-    // }
+    fn get_checkpoint<BS: Blockstore>(
+        &self,
+        store: &BS,
+        epoch: &ChainEpoch,
+    ) -> anyhow::Result<Option<Checkpoint>> {
+        let hamt = self
+            .checkpoints
+            .load(store)
+            .map_err(|e| anyhow!("failed to load checkpoints: {}", e))?;
+        let checkpoint = hamt
+            .get(&BytesKey::from(epoch.to_ne_bytes().to_vec()))
+            .map_err(|e| anyhow!("failed to get checkpoint for id {}: {:?}", epoch, e))?
+            .cloned();
+        Ok(checkpoint)
+    }
+
+    pub fn is_validator(&self, addr: &Address) -> bool {
+        self.validator_set.iter().any(|x| x.addr == *addr)
+    }
+
+    /// Do not call this function in transaction
+    pub fn verify_checkpoint<BS, RT>(&self, rt: &mut RT, ch: &Checkpoint) -> anyhow::Result<()>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        // check that subnet is active
+        if self.status != Status::Active {
+            return Err(anyhow!(
+                "submitting checkpoints is not allowed while subnet is not active"
+            ));
+        }
+
+        // check that a checkpoint for the epoch doesn't exist already.
+        if self.get_checkpoint(rt.store(), &ch.epoch())?.is_some() {
+            return Err(anyhow!("cannot submit checkpoint for epoch"));
+        };
+
+        // check that the epoch is correct
+        if ch.epoch() % self.check_period != 0 {
+            return Err(anyhow!(
+                "epoch in checkpoint doesn't correspond with a signing window"
+            ));
+        }
+
+        // check the source is correct
+        if *ch.source() != SubnetID::new(&self.parent_id, rt.message().receiver()) {
+            return Err(anyhow!("submitting checkpoint with the wrong source"));
+        }
+
+        // check previous checkpoint
+        if self.prev_checkpoint_cid(rt.store(), &ch.epoch())? != ch.prev_check().cid() {
+            return Err(anyhow!(
+                "previous checkpoint not consistent with previously committed"
+            ));
+        }
+
+        // check signature
+        let caller = rt.message().caller();
+        let pkey = self.resolve_secp_bls(rt, &caller)?;
+
+        rt.verify_signature(
+            &RawBytes::deserialize(&ch.signature().clone().into())?,
+            &pkey,
+            &ch.cid().to_bytes(),
+        )?;
+
+        Ok(())
+    }
+
+    // we need mutable reference to self due to the expected message for testing.
+    fn resolve_secp_bls<BS, RT>(&self, rt: &RT, addr: &Address) -> anyhow::Result<Address>
```

Comment:
Can we make this function part of the `fvm_uitils`? IIRC we have the exact same function in the ipc-gateway. Thanks.

Reply:
Please help check out this one: https://github.com/consensus-shipyard/fvm-utils/pull/7

Reply:
Let's use the function from that PR here once we merge it :)

Reply:
done!